### PR TITLE
feat(dummyDeployment): addition

### PR DESCRIPTION
**Pull request by eyenx**
Adds a dummyDeployment Option to be able to make the vault-secret-store provider create the secret if extraVolumes can't be used inside the Chart/App.

### CODE REVIEW

File: charts/csi-secret-provider-class/templates/_helpers.tpl [hunk 1]
```diff
@@ -0,0 +1,63 @@
+{{/* vim: set filetype=mustache: */}}
+{{/*
+Expand the name of the chart.
+*/}}
+{{- define "csi-secret-provider-class.name" -}}
```
Comment (hairmare) [inline charts/csi-secret-provider-class/templates/_helpers.tpl:5]: Most of these are covered by `common` and are not used from here.

File: charts/csi-secret-provider-class/templates/deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,46 @@
+{{- if .Values.dummyDeployment.enabled }}
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  {{ template "common.metadata" $ }}
+spec:
+  selector:
+    matchLabels:
+      {{- include "csi-secret-provider-class.selectorLabels" . | nindent 6 }}
```
Comment (hairmare) [inline charts/csi-secret-provider-class/templates/deployment.yaml:9]: There might be something generic in `common` to replace this but I'd need to check first.

File: charts/csi-secret-provider-class/values.yaml [hunk 3]
```diff
@@ -14,3 +14,34 @@ secretObjects: []
 # secretObjectsTpl -- template to generate secretObjects
 # @default -- raw values from `secretObjects`
 secretObjectsTpl: "{{ .Values.secretObjects | toYaml }}"
+
+# Creates a dummy deployment which uses the csiVolue. This is a hack, in case your Deployment in case your App/Chart isn't enabled to define extraVolumes
```
Comment (hairmare) [inline charts/csi-secret-provider-class/values.yaml:18]:
```suggestion
# Creates a dummy Deployment which uses the csiVolume. This is a hack, in case your Chart doesn't support defining extraVolumes
```

File: charts/csi-secret-provider-class/templates/deployment.yaml [hunk 4]
```diff
@@ -0,0 +1,46 @@
+{{- if .Values.dummyDeployment.enabled }}
+apiVersion: apps/v1
+kind: Deployment
```
Comment (hairmare) [inline charts/csi-secret-provider-class/templates/deployment.yaml:3]: Could this be one of the rare edge cases where we consider directly managing a Pod rather than using a Deployment?

File: charts/csi-secret-provider-class/templates/deployment.yaml (hunk 2 above)
Comment (eyenx) [reply charts/csi-secret-provider-class/templates/deployment.yaml:9]: found nothing. This is why I created `_helpers.tpl` in the first place.

File: charts/csi-secret-provider-class/templates/deployment.yaml (hunk 4 above)
Comment (eyenx) [reply charts/csi-secret-provider-class/templates/deployment.yaml:3]: We just have to think about the Deploymentstrategy. What does need to happen on changes? Remove the Pod and recreate it? The secret gets removed for a brief short of time. This is why I thought about using the Deployment, but haven't thought about Rollingupdate and so on yet.

**Comment by tongpu** [comment]
ELI5: This is required to run once to ensure that the secret is being populated? Other possible hacks I could think about is just running a Job as a Helm hook. But I'm unsure if we really need a Helm chart for everything, even hacks like that?

**Comment by eyenx** [comment]
> ELI5: This is required to run once to ensure that the secret is being populated? Other possible hacks I could think about is just running a Job as a Helm hook. But I'm unsure if we really need a Helm chart for everything, even hacks like that?

The pod needs to run always.

File: charts/csi-secret-provider-class/templates/deployment.yaml (hunk 2 above)
Comment (hairmare) [reply charts/csi-secret-provider-class/templates/deployment.yaml:9]: i took a look at the original incubator upstream as well and could not find anything there either

File: charts/csi-secret-provider-class/templates/_helpers.tpl [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+app.kubernetes.io/version: {{ .Chart.AppVersion | quote }}
+{{- end }}
+app.kubernetes.io/managed-by: {{ .Release.Service }}
+{{- end }}
+
+{{/*
+Selector labels
+*/}}
+{{- define "csi-secret-provider-class.selectorLabels" -}}
+app.kubernetes.io/name: {{ include "csi-secret-provider-class.name" . }}
```
Comment (hairmare) [inline charts/csi-secret-provider-class/templates/_helpers.tpl:50]: I think this could use something based on `common.name`.

File: charts/csi-secret-provider-class/templates/_helpers.tpl [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+*/}}
+{{- define "csi-secret-provider-class.chart" -}}
+{{- printf "%s-%s" .Chart.Name .Chart.Version | replace "+" "_" | trunc 63 | trimSuffix "-" }}
+{{- end }}
+
+{{/*
+Common labels
+*/}}
+{{- define "csi-secret-provider-class.labels" -}}
+helm.sh/chart: {{ include "csi-secret-provider-class.chart" . }}
```
Comment (hairmare) [inline charts/csi-secret-provider-class/templates/_helpers.tpl:38]: Might be replaceable with `common.chartref`

File: charts/csi-secret-provider-class/templates/_helpers.tpl [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+*/}}
+{{- define "csi-secret-provider-class.selectorLabels" -}}
+app.kubernetes.io/name: {{ include "csi-secret-provider-class.name" . }}
+app.kubernetes.io/instance: {{ .Release.Name }}
+{{- end }}
+
+{{/*
+Create the name of the service account to use
+*/}}
+{{- define "csi-secret-provider-class.serviceAccountName" -}}
```
Comment (hairmare) [inline charts/csi-secret-provider-class/templates/_helpers.tpl:57]: This one isn't used anywhere

**Review by hairmare — COMMENTED** [review]
some more food for thought regarding using common in helpers

File: charts/csi-secret-provider-class/templates/deployment.yaml (hunk 4 above)
Comment (hairmare) [reply charts/csi-secret-provider-class/templates/deployment.yaml:3]: You are right, the changes will be much more straight forward if it's a Deployment.

**Comment by hairmare** [comment]
> This is required to run once to ensure that the secret is being populated? Other possible hacks I could think about is just running a Job as a Helm hook

A "run once" solution might be done with [argocd-vault-plugin](https://github.com/IBM/argocd-vault-plugin) but this aims at tricking the csi bits into continually managing the secret.

File: charts/csi-secret-provider-class/values.yaml [hunk 8]
```diff
@@ -14,3 +14,34 @@ secretObjects: []
 # secretObjectsTpl -- template to generate secretObjects
 # @default -- raw values from `secretObjects`
 secretObjectsTpl: "{{ .Values.secretObjects | toYaml }}"
+
+# Creates a dummy Deployment which uses the csiVolume. This is a hack, in case your Chart doesn't support defining extraVolumes
```
Comment (hairmare) [inline charts/csi-secret-provider-class/values.yaml:18]: Maybe we should add another strongly worded sentence here that tells people not to use this unless they tried fixing the underlying issue and can prove this with a PR or support ticket for the affected chart.

File: charts/csi-secret-provider-class/templates/_helpers.tpl (hunk 7 above)
Comment (eyenx) [reply charts/csi-secret-provider-class/templates/_helpers.tpl]: 11d0f3e759ba214811f4d382bfbac26083874c9b

File: charts/csi-secret-provider-class/templates/_helpers.tpl (hunk 5 above)
Comment (eyenx) [reply charts/csi-secret-provider-class/templates/_helpers.tpl]: 11d0f3e759ba214811f4d382bfbac26083874c9b

File: charts/csi-secret-provider-class/templates/_helpers.tpl (hunk 6 above)
Comment (eyenx) [reply charts/csi-secret-provider-class/templates/_helpers.tpl]: 11d0f3e759ba214811f4d382bfbac26083874c9b

File: charts/csi-secret-provider-class/templates/_helpers.tpl (hunk 1 above)
Comment (eyenx) [reply charts/csi-secret-provider-class/templates/_helpers.tpl]: 11d0f3e759ba214811f4d382bfbac26083874c9b

File: charts/csi-secret-provider-class/templates/_helpers.tpl (hunk 6 above)
Comment (eyenx) [reply charts/csi-secret-provider-class/templates/_helpers.tpl]: 11d0f3e759ba214811f4d382bfbac26083874c9b

File: charts/csi-secret-provider-class/values.yaml (hunk 8 above)
Comment (eyenx) [reply charts/csi-secret-provider-class/values.yaml:18]:
```suggestion
# Creates a dummy Deployment which uses the csiVolume. This is a hack, in case your Chart doesn't support defining extraVolumes
# DO NOT USE THIS unless strictly needed. Always try to use extraVolume Bindings inside your already existing Chart or application to make csi-secret-store manage the secret
```

Comment (eyenx) [reply charts/csi-secret-provider-class/values.yaml:18]: Something like this?

File: charts/csi-secret-provider-class/values.yaml [hunk 9]
```diff
@@ -14,3 +14,34 @@ secretObjects: []
...
+  enabled: false
+  image:
+    # dummyDeployment.image.repository -- Container image to deploy
+    # @default -- k8s.gcr.io/pause
+    repository: k8s.gcr.io/pause
+    # dummyDeployment.image.tag -- sets the image tag to use
+    # @default -- 3.4.1
+    tag: 3.4.1
+    # dummyDeployment.image.pullPolicy -- When to pull the container image
+    # @defualt -- ifNotPresent
```
Comment (tongpu) [inline charts/csi-secret-provider-class/values.yaml:31]:
```suggestion
    # @default -- ifNotPresent
```

File: charts/csi-secret-provider-class/templates/deployment.yaml [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+      {{- with .Values.dummyDeployment.affinity }}
+      affinity:
+        {{- toYaml . | nindent 8 }}
+      {{- end }}
+      {{- with .Values.dummyDeployment.tolerations }}
+      tolerations:
+        {{- toYaml . | nindent 8 }}
+      {{- end }}
+      volumes:
+        - name: csi-secret-store-volume
```
Comment (tongpu) [inline charts/csi-secret-provider-class/templates/deployment.yaml:40]: Is it sufficient to define the volume in `volumes` without defining a `volumeMount` in any container?

**Review by tongpu — COMMENTED** [review]
This discussion is probably beyond the scope of the pull request, but what would happen if this deployment is never scheduled because of resource exhaustion?
The secret would never be created and the other application without the `extraVolumes` will also never be scheduled, because it's waiting for a secret to be available in the Kubernetes API?

**Comment by hairmare** [comment]
> This discussion is probably beyond the scope of the pull request, but what would happen if this deployment is never scheduled because of resource exhaustion?

It's a tiny pause container, if we run into this then we have large issues elsewhere. We should probably test how it reacts in those scenarios at some point though. I would not be surprised if we find room for improvement in these areas of the CSI stack.

File: charts/csi-secret-provider-class/templates/deployment.yaml (hunk 10 above)
Comment (eyenx) [reply charts/csi-secret-provider-class/templates/deployment.yaml:40]: yes.

**Comment by eyenx** [comment]
> > This is required to run once to ensure that the secret is being populated? Other possible hacks I could think about is just running a Job as a Helm hook
> 
> A "run once" solution might be done with [argocd-vault-plugin](https://github.com/IBM/argocd-vault-plugin) but this aims at tricking the csi bits into continually managing the secret.

lol we might keep an eye open for that plugin.

File: charts/csi-secret-provider-class/values.yaml (hunk 8 above)
Comment (hairmare) [reply charts/csi-secret-provider-class/values.yaml:18]: yeah & with an additional mention of the upstream work that we expect you to put in before you are allowed to use it

File: charts/csi-secret-provider-class/templates/deployment.yaml (hunk 10 above)
Comment (hairmare) [reply charts/csi-secret-provider-class/templates/deployment.yaml:40]: s/yes/maybe/ 🤨

Comment (eyenx) [reply charts/csi-secret-provider-class/templates/deployment.yaml:40]: AH!

Comment (eyenx) [reply charts/csi-secret-provider-class/templates/deployment.yaml:40]: But not by the "dummyDeployment" which is NOT using the Secret right away.

File: charts/csi-secret-provider-class/values.yaml (hunk 8 above)
Comment (eyenx) [reply charts/csi-secret-provider-class/values.yaml:18]: Better?

Comment (hairmare) [reply charts/csi-secret-provider-class/values.yaml:18]: yes, thanks!

File: charts/csi-secret-provider-class/templates/deployment.yaml (hunk 10 above)
Comment (hairmare) [reply charts/csi-secret-provider-class/templates/deployment.yaml:40]: we might want to investigate this further, i can't really explain what has been happeining yet but i can take another deep dive into the implementation to figure it out if we don't identify a pattern.

File: charts/csi-secret-provider-class/values.yaml (hunk 8 above)
Comment (hairmare) [reply charts/csi-secret-provider-class/values.yaml:18]: Did you end up changing this?

Comment (eyenx) [reply charts/csi-secret-provider-class/values.yaml:18]: Just now.

Comment (eyenx) [reply charts/csi-secret-provider-class/values.yaml:18]: Now checks are failing... ngg

File: charts/csi-secret-provider-class/values.yaml [hunk 11]
```diff
@@ -14,3 +14,36 @@ secretObjects: []
 # secretObjectsTpl -- template to generate secretObjects
 # @default -- raw values from `secretObjects`
 secretObjectsTpl: "{{ .Values.secretObjects | toYaml }}"
+
+# Creates a dummy Deployment which uses the csiVolume. This is a hack, in case your Chart doesn't support defining extraVolumes
+# DISCLAIMER: DO NOT USE THIS unless as last resort.
```
Comment (eyenx) [inline charts/csi-secret-provider-class/values.yaml:19]:
```suggestion
# DISCLAIMER: DO NOT USE THIS unless as last resort.
```

File: charts/csi-secret-provider-class/values.yaml (hunk 8 above)
Comment (eyenx) [reply charts/csi-secret-provider-class/values.yaml:18]: Checks are passing @hairmare

File: charts/csi-secret-provider-class/templates/deployment.yaml [hunk 12]
```diff
@@ -0,0 +1,46 @@
+{{- if .Values.dummyDeployment.enabled }}
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  {{ template "common.metadata" $ }}
```
Comment (hairmare) [inline charts/csi-secret-provider-class/templates/deployment.yaml:5]: The nesting seems wrong here:
```suggestion
{{ template "common.metadata" . }}
```
Also, there is no reason to pass `$` when `.` contains the same content.

**Review by hairmare — CHANGES_REQUESTED** [review]
Apart from the nesting error below you should also add `dummyDeployment.enabled=true` to `ci/default-values.yaml` for ensuring that the test generates a deployment.

File: charts/csi-secret-provider-class/templates/deployment.yaml [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+        {{- include "csi-secret-provider-class.selectorLabels" . | nindent 8 }}
+    spec:
+      containers:
+        - name: dummy
+          image: {{ .Values.dummyDeployment.image.repository }}:{{ .Values.dummyDeployment.image.tag }}
+          imagePullPolicy: {{ .Values.dummyDeployment.image.pullPolicy }}
+          {{- with .Values.dummyDeployment.resources }}
+          resources:
+            {{- toYaml .Values.dummyDeployment.resources | nindent 12 }}
+          {{- end }}
```
Comment (hairmare) [inline charts/csi-secret-provider-class/templates/deployment.yaml:25]: Use `.` when using `with`.
```suggestion
          {{- with .Values.dummyDeployment.resources }}
          resources:
            {{- toYaml .| nindent 12 }}
          {{- end }}
```

File: charts/csi-secret-provider-class/README.md [hunk 14]
```diff
@@ -17,12 +17,21 @@ This chart is maintained by [Adfinis](https://adfinis.com/?pk_campaign=github&pk
...
-| https://charts.adfinis.com | common | 0.0.5 |
+| https://charts.adfinis.com | common | 0.0.6 |
 
 ## Values
 
 | Key | Type | Default | Description |
 |-----|------|---------|-------------|
+| dummyDeployment.affinity | object | {} | sets affinity rules on the  deployment |
+| dummyDeployment.enabled | bool | false | enable the dummy deployment |
+| dummyDeployment.image.pullPolicy | string | ifNotPresent | When to pull the container image |
```
Comment (hairmare) [inline charts/csi-secret-provider-class/README.md:28]:
```suggestion
| dummyDeployment.image.pullPolicy | string | IfNotPresent | When to pull the container image |
```

File: charts/csi-secret-provider-class/values.yaml [hunk 15]
```diff
@@ -14,3 +14,36 @@ secretObjects: []
...
+  image:
+    # dummyDeployment.image.repository -- Container image to deploy
+    # @default -- k8s.gcr.io/pause
+    repository: k8s.gcr.io/pause
+    # dummyDeployment.image.tag -- sets the image tag to use
+    # @default -- 3.4.1
+    tag: 3.4.1
+    # dummyDeployment.image.pullPolicy -- When to pull the container image
+    # @default -- ifNotPresent
+    pullPolicy: ifNotPresent
```
Comment (hairmare) [inline charts/csi-secret-provider-class/values.yaml:34]:
```suggestion
    pullPolicy: IfNotPresent
```
Also, don't duplicate `@default` if it matches the actual value!

**Review by hairmare — CHANGES_REQUESTED** [review]
Found the typo!

File: charts/csi-secret-provider-class/values.yaml [hunk 16]
```diff
@@ -14,3 +14,36 @@ secretObjects: []
...
+  # dummyDeployment.enabled -- enable the dummy deployment
+  # @default -- false
+  enabled: false
+  image:
+    # dummyDeployment.image.repository -- Container image to deploy
+    # @default -- k8s.gcr.io/pause
+    repository: k8s.gcr.io/pause
+    # dummyDeployment.image.tag -- sets the image tag to use
+    # @default -- 3.4.1
+    tag: 3.4.1
```
Comment (hairmare) [inline charts/csi-secret-provider-class/values.yaml:31]:
```suggestion
  enabled: false
  image:
    # dummyDeployment.image.repository -- Container image to deploy
    repository: k8s.gcr.io/pause
    # dummyDeployment.image.tag -- sets the image tag to use
    tag: 3.4.1
```

File: charts/csi-secret-provider-class/values.yaml [hunk 17]
```diff
@@ -14,3 +14,36 @@ secretObjects: []
...
+  nodeSelector: {}
+  # dummyDeployment.affinity -- sets affinity rules on the  deployment
+  # @default -- {}
+  affinity: {}
+  # dummyDeployment.tolerations -- sets tolerations on the  deployment
+  # @default -- []
+  tolerations: []
+  # dummyDeployment.resources -- sets resources like limits and requests on the  deployment
+  # @default -- {}
+  resources: {}
```
Comment (hairmare) [inline charts/csi-secret-provider-class/values.yaml:49]:
```suggestion
  podAnnotations: {}
  # dummyDeployment.nodeSelector -- sets a nodeSelector on the  deployment
  nodeSelector: {}
  # dummyDeployment.affinity -- sets affinity rules on the  deployment
  affinity: {}
  # dummyDeployment.tolerations -- sets tolerations on the  deployment
  tolerations: []
  # dummyDeployment.resources -- sets resources like limits and requests on the  deployment
  resources: {}
```

**Review by hairmare — CHANGES_REQUESTED** [review]
Only use `@default` when it you need to override the generated default because it would produce illegible JSON output.

**Review by hairmare — APPROVED** [review]
